numpy/lib/arraypad.py

```diff
@@ -69,821 +81,475 @@ def _round_ifneeded(arr, dtype):
+        padded.fill(fill_value)
 
-    # Generate an n-dimensional array incrementing along `axis`
-    ramp_arr = _arange_ndarray(arr, padshape, axis,
-                               reverse=False).astype(np.float64)
+    # Copy old array into correct space
+    original_area_slice = tuple(
+        slice(left, left + size)
+        for size, (left, right) in zip(array.shape, pad_width)
+    )
+    padded[original_area_slice] = array
 
-    # Slice a chunk from the edge to calculate stats on
-    edge_slice = _slice_last(arr.shape, 1, axis=axis)
+    return padded, original_area_slice
 
-    # Extract edge, and extend along `axis`
-    edge_pad = arr[edge_slice].repeat(pad_amt, axis)
 
-    # Linear ramp
-    slope = (end - edge_pad) / float(pad_amt)
-    ramp_arr = ramp_arr * slope
-    ramp_arr += edge_pad
-    _round_ifneeded(ramp_arr, arr.dtype)
-
-    # Ramp values will most likely be float, cast them to the same type as arr
-    return _do_append(arr, ramp_arr, axis)
-
-
-def _prepend_max(arr, pad_amt, num, axis=-1):
+def _set_pad_area(padded, axis, index_pair, value_pair):
     """
-    Prepend `pad_amt` maximum values along `axis`.
+    Set empty-padded area in given dimension.
 
     Parameters
     ----------
-    arr : ndarray
-        Input array of arbitrary shape.
-    pad_amt : int
-        Amount of padding to prepend.
-    num : int
-        Depth into `arr` along `axis` to calculate maximum.
-        Range: [1, `arr.shape[axis]`] or None (entire axis)
+    padded : ndarray
+        Array with the pad area which is modified inplace.
     axis : int
-        Axis along which to pad `arr`.
-
-    Returns
-    -------
-    padarr : ndarray
-        Output array, with `pad_amt` values appended along `axis`. The
-        prepended region is the maximum of the first `num` values along
-        `axis`.
-
+        Dimension with the pad area to set.
+    index_pair : (int, int)
+        Pair of indices that mark the end (or start) of the pad area on both
+        sides in the given dimension.
+    value_pair : tuple of scalars or ndarrays
+        Values inserted into the pad area on each side. It must match or be
+        broadcastable to the shape of `arr`.
     """
-    if pad_amt == 0:
-        return arr
-
-    # Equivalent to edge padding for single value, so do that instead
-    if num == 1:
-        return _prepend_edge(arr, pad_amt, axis)
-
-    # Use entire array if `num` is too large
-    if num is not None:
-        if num >= arr.shape[axis]:
-            num = None
-
-    # Slice a chunk from the edge to calculate stats on
-    max_slice = _slice_first(arr.shape, num, axis=axis)
+    if index_pair[0] > 0:
+        # Set pad values on the left
+        left_slice = _slice_at_axis(
+            padded.shape, slice(None, index_pair[0]), axis)
+        padded[left_slice] = value_pair[0]
 
-    # Extract slice, calculate max
-    max_chunk = arr[max_slice].max(axis=axis, keepdims=True)
+    if index_pair[1] > 0:
```

Comment:
There may be funny things possible due to `_as_pairs` casting all to one dtype as well, but nothing changed there, so its fine.

Reply:
> OK, so this if here is only necessary, because you use -0 below otherwise?

Yes.

> Now np.pad(np.ones(1), (1, 0), constant_values=(1, "asdf")) stopped raising an error.

When did it raise an error? Numpy 1.15.1 will return `array([1., 1.])`. As does my code.

> This means that you need to fix up the linear ramp, but I think that is a good idea in any case.

I introduced a sanity check for `size=0` in 71bc288. Is that what you mean?

> Sorry, probably I have to think about it a bit more. Sometimes padding may not be wanted at all, and 0 is just used to indicate no padding, so that errors may not be wanted.

Saw this to late. This may be a good reason to keep the `if width_pair[...] > 0` statements (`width_pair` was `index_pair`) at least for this function. For now I've done away with all the special treatment when the `pad_width` is 0. Let me know if that is what you had in mind.

Keeping track of the feedback sometimes gets a little bit confusing. :sweat_smile:

Reply:
Hmmm, not sure where I saw that error, maybe I mixed something up, or was using an older python version on a different computer at the time. No error is good in any case. A test would be good.

Doing away with the "0" paths was what I thought. I just was not sure if some tricky things might crop up. I guess it might micro-optimize 0 pad width for small arrays a tiny bit if the `if` is added, but I do not think that is worth extra code paths.